Update to iOS 13

angstsmurf opened the pull request:
I'd thought I'd make a pull request just in case there is still any interest in this.

To actually build and use it you will need the separate "Update to iOS 13" PRs for IosGlulxe and Glulxe, or the corresponding PRs for IosFizmo and Fizmo.

I’ve fixed most of the low-hanging fruit, but there is still much to do.

~~I converted the project to automatic reference counting and it seems to work fine, but I suppose what you wrote about the interpreter core not being compatible means that we leak a lot of memory. I haven't looked into this. The Leaks tool crashes as soon as the interpreter starts running, which is a clear sign that something isn't right.~~

The minimum deployment is set to iOS 13.0 (in IosGlulxe and IosFizmo, not the IosGlk project file, which I haven’t touched at all.) This seemed like a reasonable bump, but likely means that people on older systems and device won’t be able to download it from App Store anymore.

~~I have implemented Dark Mode support, but the Notes view is still black on bright yellow in Dark Mode, and also the Help and License web views are unchanged. Ideally the game view should also switch to a dark mode automatically.~~

~~The line spacing in the Notes view is off. We should draw the lines on the fly instead after measuring the actual spacing of the text.~~

~~Remaining to do is updating the text system to support the standard VoiceOver rotor and so on.~~ And then converting it all to Swift and Swift UI, of course.


EDIT: It kind of works now, using UITextViews for the text, but it isn't rock-solid. ~~The main thing that needs fixing is state restoration. If you background the app while the Notes, Help or Settings tab is active rather than the game tab, the game view will be scrolled to the top with a More icon showing when you return to it. (Also, it will return to the game tab instead of whatever was showing when you backgrounded it.)~~ EDIT: Fixed.

There are still minor bugs and glitches related to the fact that the frame view changes height every time the keyboard shows or hides, and it shows and hides a lot, so anything that relies on having the correct measurements gets confused.

angstsmurf committed:
Convert to automatic reference counting

angstsmurf committed:
Remove deprecated stuff

angstsmurf committed:
Remove deprecated stuff

angstsmurf committed:
Preliminary autorestore fix

angstsmurf committed:
Auto-modernize syntax

angstsmurf committed:
Initial implementation of storyboards

angstsmurf committed:
Fix alerts on iPad

angstsmurf committed:
Color fixes

Foreground color was not honoured when drawing letters.
Textview background was not updated after a change, which could
be seen as lines of the old color at the top and bottom, especially when
scrolling.

angstsmurf committed:
Silence analyzer

angstsmurf committed:
Use weak references to self in blocks

angstsmurf committed:
Fix navigating away from file prompts on Ipad

On Ipad, alerts that are modal on Iphone can be dismissed by swiping.
This would just hang the game.

angstsmurf committed:
Adjust Notes view for iPad

angstsmurf committed:
Set deployment to iOS 13 and remove deprecations

Also fix some animation coordinates.

angstsmurf committed:
Let's try keeping the GlkFileSelect storyboard here instead

For some reason I couldn't get Fizmo to compile otherwise.

erkyrath commented:
You're welcome to work on this in a fork, but I'm not accepting PRs for this.

> I converted the project to automatic reference counting and it seems to work fine, but I suppose what you wrote about the interpreter core not being compatible means that we leak a lot of memory.

That is what I expect.

The critical bit is in GlkAppWrapper.m, where we "drain and recreate the thread's autorelease pool". This happens in every glk_select() call, which is good time to clear out dead memory. However -- if I recall correctly -- ARC doesn't allow you to do this inside a function. You have to put your code inside a `@autoreleasepool {}` block, which releases dead memory when the block exits.

Since the entire game runs in a non-exiting loop, there's nowhere to put the `@autoreleasepool {}` block.

angstsmurf committed:
Half-assed modernisations

Mainly eliminating @synthesize.

angstsmurf committed:
Disable ARC for GlkAppWrapper.m only

angstsmurf commented:
Right, thanks for pointing this out. We'll just have to leave ARC off on the GlkAppWrapper.m file only for now, while keeping it on for the rest.

It is annoying, because most objects do get deallocated when they should with ARC, it's just a couple of them that refuse, even though they don't seem to be referenced by anything other than the autoreleasepool object.

By the way, did you ever give any thought to how one would restructure the VM loop to allow for ARC, given infinite amounts of time and motivation?

erkyrath commented:
> We'll just have to leave ARC off on the GlkAppWrapper.m file only for now, while keeping it on for the rest.

I'm not sure that works. I suspect (without digging into it at all) that for ARC files, allocations are managed by a statically-scoped release pool, not by the one that gets drained in GlkAppWrapper.m.

> By the way, did you ever give any thought to how one would restructure the VM loop to allow for ARC, given infinite amounts of time and motivation?

I think you'd have to compile the whole Glulxe interpreter as ObjC with ARC on. Then turn execute_loop() into a double loop: one loop nested inside `@autoreleasepool` inside another loop. The inner one would return periodically (after every select() call).

It's more of a nuisance in Swift, since Swift isn't meant to compile C code as-is.

angstsmurf committed:
Get rid of more @synthesize

angstsmurf committed:
Try to fix ARC by compiling Glulxe as ObjC

Compile the whole Glulxe interpreter as ObjC with ARC.
Turn execute_loop() into a double loop: one loop nested inside @autoreleasepool inside another loop.
The inner one returns periodically (after every select() call).

angstsmurf commented:
Yes, that seems to work. Thanks again!

angstsmurf committed:
Delete leftover comment

angstsmurf committed:
System dark theme makes interpreter dark as well

We remove the dark option from the color settings menu,
for the sake of simplicity.

angstsmurf committed:
Initial update of text system

Removed GlkStyledLine and GlkVisualLine and replaced them with UITextViews.

angstsmurf committed:
Add custom rotors

angstsmurf committed:
Create rotors in IosGlkViewController viewWillAppear

angstsmurf committed:
WIP text system

Lots of problems still

angstsmurf committed:
Fixes for gridview

Mostly by trial and error and comparison to the old version.
Not sure if input field placement is correct.

angstsmurf committed:
Cleanup

angstsmurf committed:
GlkFrameView: Add fixme comment about menu arrange event bug

angstsmurf committed:
Clean up more unused stuff

Such as the TextSelectView class

angstsmurf committed:
Let's just commit the rest of the WIP stuff

angstsmurf committed:
Scrolling fixes

angstsmurf committed:
Clean up debug NSLog spam

angstsmurf committed:
More cleanup

angstsmurf committed:
More scrolling fixes

angstsmurf committed:
Hacky attempt to fix more symbol and speech at startup

angstsmurf committed:
More cleanup

angstsmurf committed:
Delete unused files

angstsmurf committed:
Preliminary state restoration fixes and cleanup

angstsmurf committed:
Clean up GlkWindowViewUIState

It should probably be an extension instead.

angstsmurf committed:
Create a terp subclass of the IosGlkTabBarControllerDelegate

angstsmurf committed:
Rename method

angstsmurf committed:
Clean up

angstsmurf committed:
Update headers for the new files

angstsmurf committed:
Remove unused property tagobj

angstsmurf commented:
It seems to pretty much work now. Ready for testing.
